designer/server/src/main/scala/pl/touk/nussknacker/ui/process/deployment/DeploymentServiceImpl.scala

```diff
@@ -240,73 +249,118 @@ class DeploymentServiceImpl(
   }
 
   private def checkIfCanPerformActionInState[PS: ScenarioShapeFetchStrategy](
-      actionType: ProcessActionType,
+      actionName: ScenarioActionName,
       processDetails: ScenarioWithDetailsEntity[PS],
       ps: ProcessState
-  ): Unit = {
-    if (!ps.allowedActions.contains(actionType)) {
-      logger.debug(s"Action: $actionType on process: ${processDetails.name} not allowed in ${ps.status} state")
-      throw ProcessIllegalAction(ScenarioActionName(actionType), processDetails.name, ps)
+  )(implicit user: LoggedUser): Unit = {
+    val allowedActions = allowedActionsForState(processDetails, ps)
+    if (!allowedActions.contains(actionName)) {
+      logger.debug(s"Action: $actionName on process: ${processDetails.name} not allowed in ${ps.status} state")
+      throw ProcessIllegalAction(actionName, processDetails.name, ps.status.name, allowedActions)
     }
   }
 
-  protected def prepareDeploymentData(user: User, deploymentId: DeploymentId): DeploymentData = {
-    DeploymentData(deploymentId, user, Map.empty)
+  private def allowedActionsForState[PS: ScenarioShapeFetchStrategy](
+      processDetails: ScenarioWithDetailsEntity[PS],
+      ps: ProcessState
+  )(implicit user: LoggedUser): List[ScenarioActionName] = {
+    val defaultStateActions = ps.allowedActions.map(ScenarioActionName(_))
```

Comment:
Why `default`? WDYT about `actionsDefinedInState` and `actionsDefinedInCustomActions`?

Reply:
Fixed.